### PR TITLE
Handle kotlin compiler args in dev mode

### LINKED ISSUE #15418: Dev reload ignores Kotlin compiler arguments
**Describe the bug**
When using Kotlin's `Result` type (which is somewhat experimental), you must pass an argument to the compiler to enable Result type to be returned by a function.

Argument required to enable Result return type: `-Xallow-result-return-type`

<details>
<summary>How to pass args to build tools</summary>
<br>

If using maven:

```xml
<plugin>
    <artifactId>kotlin-maven-plugin</artifactId>
    <configuration>
        <jvmTarget>1.8</jvmTarget>
        <args>
            <arg>-Xallow-result-return-type</arg>
        </args>
    </configuration>
    <groupId>org.jetbrains.kotlin</groupId>
    <version>${kotlin.version}</version>
```

If using gradle:

```gradle
compileKotlin {
    kotlinOptions.jvmTarget = "1.8"
    kotlinOptions.freeCompilerArgs = ["-Xallow-result-return-type"]

}
compileTestKotlin {
    kotlinOptions.jvmTarget = "1.8"
    kotlinOptions.freeCompilerArgs = ["-Xallow-result-return-type"]
}
```
</details>

Writing a simple hello world example using the Result type works on first compilation:

```kt
@Path("/hello")
class GreetingResource {

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    fun hello4() = getAGreetingFailure().getOrDefault("Something went wrong")

    fun getAGreetingFailure(): Result<String> = kotlin.runCatching {
        throw Error("An Error")
    }
}
```

**Expected behavior**

- When running in dev mode 
- When changing the string inside error (`Error("Something else")`) or making any other trivial change

Quarkus is able to "reload" the application and continue working as expected.

**Actual behavior**
(Describe the actual behavior clearly and concisely.)

On reload, instead of getting back `"Something else"` when hitting the `/hello` endpoint, I see compilation errors `Kotlin.Result` cannot be used as a return type

image

**To Reproduce**

Link to a small reproducer - https://github.com/MrMoner/kotlin-result-pattern-demo

1. Hit localhost /hello in dev mode
2. Make a trivial change to force reload
3. Hit endpoint again

**Environment (please complete the following information):**
 - Output of `ver`:  Microsoft Windows [Version 10.0.19041.844]
 - Output of `java -version`:  
openjdk version "11.0.10" 2021-01-19
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.10+9)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.10+9, mixed mode)
 - Quarkus version: 1.12.0.Final
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): apache-maven-3.6.3

### PR DESCRIPTION
**Pull request by glefloch**
This sets compilerArgs from `DevMojo`/`QuarkusDev` tasks in the `KotlinCompilationProvider`.
This allows a user to propagate kotlin compilation arguments in the quarkus plugin for devmode. 

I created this as a draft PR because I encounter a weird classloader issue :

```
java.lang.ClassCastException: class kotlin.jvm.internal.ClassReference cannot be cast to class kotlin.reflect.jvm.internal.KClassImpl (kotlin.jvm.internal.ClassReference is in unnamed module of loader 'app'; kotlin.reflect.jvm.internal.KClassImpl is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @6321e813)
	at kotlin.reflect.full.KClasses.getMemberProperties(KClasses.kt:149)
	at org.jetbrains.kotlin.cli.common.arguments.ParseCommandLineArgumentsKt.parsePreprocessedCommandLineArguments(parseCommandLineArguments.kt:74)
	at org.jetbrains.kotlin.cli.common.arguments.ParseCommandLineArgumentsKt.parseCommandLineArguments(parseCommandLineArguments.kt:67)
	at org.jetbrains.kotlin.cli.common.CLITool.parseArguments(CLITool.kt:118)
	at io.quarkus.kotlin.deployment.KotlinCompilationProvider.compile(KotlinCompilationProvider.java:78)
	at io.quarkus.deployment.dev.ClassLoaderCompiler.compile(ClassLoaderCompiler.java:176)
	at io.quarkus.deployment.dev.RuntimeUpdatesProcessor.checkForChangedClasses(RuntimeUpdatesProcessor.java:352)
	at io.quarkus.deployment.dev.RuntimeUpdatesProcessor.doScan(RuntimeUpdatesProcessor.java:195)
	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$2.handle(VertxHttpHotReplacementSetup.java:62)
	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$2.handle(VertxHttpHotReplacementSetup.java:52)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:313)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

@geoand do you have any pointers? 

close #15418

### CODE REVIEW

**Comment by glefloch** [comment]
yes, that what I thought, there is already most of kotlin dependencies declared as parent-first, I will try to spot which one is missing

**Comment by gsmet** [comment]
@glefloch do you think you could finalize this one soon? (trying to converge on CR1 I will release tomorrow)

**Comment by glefloch** [comment]
@gsmet I should be able to work on this today, it should be ok.

**Comment by glefloch** [comment]
I tried to add/remove most of the kotlin library in parent-first but I was not able to find a working combination. 
Can it be something linked to the dev mode class loader?  maybe @stuartwdouglas have an idea? 

I also tried to reimplement the function in java, I can list all fields by reflection (as done by the method we are trying to call) but as all fields are delegated a class, I'm not able to get their annotations. (and I don't think this is a good option..)

**Comment by stuartwdouglas** [comment]
How can I reproduce this issue?

**Comment by glefloch** [comment]
This issue contains a simple reproducer. If you build my branch and try to run it in quarkus devmode you will get this error as soon as the application restart after a change.

**Comment by stuartwdouglas** [comment]
I just tried this rebased on master, and I don't see any class loading issues, but I don't see anything actually setting the kotlin compiler args. Shouldn't this also need a change in QuarkusDev/DevMojo to actually extract the kotlin arguments?

Also what would happen if both Java and kotlin where present?

**Comment by glefloch** [comment]
I just did the same test and still have the issue. Actually, maven and gradle already handle extra compiler args, they were just not set in the kotlin compiler. 
They are handled here: https://github.com/quarkusio/quarkus/pull/15552/files#diff-0eeb57213e1a5f6311aa8310faf427d33ca34e303566c17ee5000a9610d056baR78

Quarkus dev mode starts correctly but as soon as I update the `GreetingResource.kt` and re-hit the endpoint, I get the following error: 
````
java.lang.ClassCastException: class kotlin.jvm.internal.ClassReference cannot be cast to class kotlin.reflect.jvm.internal.KClassImpl (kotlin.jvm.internal.ClassReference is in unnamed module of loader 'app'; kotlin.reflect.jvm.internal.KClassImpl is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @7892644b)
	at kotlin.reflect.full.KClasses.getMemberProperties(KClasses.kt:149)
	at org.jetbrains.kotlin.cli.common.arguments.ParseCommandLineArgumentsKt.parsePreprocessedCommandLineArguments(parseCommandLineArguments.kt:74)
	at org.jetbrains.kotlin.cli.common.arguments.ParseCommandLineArgumentsKt.parseCommandLineArguments(parseCommandLineArguments.kt:67)
	at org.jetbrains.kotlin.cli.common.CLITool.parseArguments(CLITool.kt:118)
	at io.quarkus.kotlin.deployment.KotlinCompilationProvider.compile(KotlinCompilationProvider.java:78)
````

Regarding multi source project, that's a really good question, compiler args are collected and set in the context but I don't think there is any way to differentiate them.

**Comment by glefloch** [comment]
Ok, I think the reproducer is not ok, here is mine:
[15552.zip](https://github.com/quarkusio/quarkus/files/6491804/15552.zip)

**Comment by stuartwdouglas** [comment]
I have rebased and added a fix to the CL issue.

**Comment by glefloch** [comment]
Thanks a lot @stuartwdouglas !

**Comment by stuartwdouglas** [comment]
I think there may be some issues with the test though, I only looked at the ClassCast issue.

**Comment by glefloch** [comment]
Yes I wasn't able to run it. I will take care of fixing it.

**Comment by glefloch** [comment]
@stuartwdouglas, I just tested it but I still have the `ClassCastException` and I can't see your modification.

**Comment by stuartwdouglas** [comment]
Did you pull from your fork? The fix is to add kotlin-reflect to the kotlin
runtime module. Parent first only works if the artifact is in the parent
cl, and in this case kotlin-reflect was only a deployment dep so the Dev
mojo was not adding it to the class path.

**Comment by glefloch** [comment]
CI looks good, sorry for the delay.

**Comment by glefloch** [comment]
@geoand, @evanchooly I you want to have a look too ?

File: extensions/kotlin/runtime/pom.xml [hunk 1]
```diff
@@ -56,6 +56,10 @@
             <groupId>org.jetbrains.kotlin</groupId>
             <artifactId>kotlin-stdlib-jdk8</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.jetbrains.kotlin</groupId>
+            <artifactId>kotlin-reflect</artifactId>
+        </dependency>
```
Comment (geoand) [inline extensions/kotlin/runtime/pom.xml:62]: Is this now needed?

Comment (glefloch) [reply extensions/kotlin/runtime/pom.xml:62]: Yes this is the fix Stuart added. If the dependency is not part the runtime artifact, then parentfirst does not work and we end up we a class cast exception.

Comment (geoand) [reply extensions/kotlin/runtime/pom.xml:62]: Understood, thanks!

File: integration-tests/kotlin/src/test/resources/projects/kotlin-compiler-args/src/main/resources/META-INF/resources/index.html [hunk 2]
```diff
@@ -0,0 +1,156 @@
+<!DOCTYPE html>
```
Comment (geoand) [inline integration-tests/kotlin/src/test/resources/projects/kotlin-compiler-args/src/main/resources/META-INF/resources/index.html:1]: Can we get rid of this, it's not really needed

**Review by geoand — APPROVED** [review]

**Comment by glefloch** [comment]
@geoand I pushed the `META-INF/resources/index.html` back, it is needed to check the dev mode started here: https://github.com/quarkusio/quarkus/pull/15552/files#diff-41d0b36caccbfc00e6e30fdd6803f0d732400fa5f58b56dd6ee37a08c80cabbdR64

**Comment by geoand** [comment]
Aha, gotcha

**Review by evanchooly — APPROVED** [review]
looks pretty straightforward to me.

File: integration-tests/kotlin/src/test/resources/projects/kotlin-compiler-args/pom.xml [hunk 3]
```diff
@@ -0,0 +1,133 @@
+<?xml version="1.0"?>
+<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>org.acme</groupId>
+  <artifactId>acme</artifactId>
+  <version>1.0.0-SNAPSHOT</version>
+  <properties>
+    <compiler-plugin.version>3.8.1</compiler-plugin.version>
+    <kotlin.version>1.4.20</kotlin.version>
```
[the remark this answers is not in the document]
Comment (glefloch) [reply integration-tests/kotlin/src/test/resources/projects/kotlin-compiler-args/pom.xml]: I will go through all kotlin extension to sync this version.

**Comment by glefloch** [comment]
Can I merge this one @geoand ?

**Comment by geoand** [comment]
Yes certainly!

Thanks for the work!

**Pull request merged**